Added more sensors

matissrapsa opened the pull request:
Added more sensors for version 1.0.3
Oscillation Range, Oscillation Speed, Vibration Strength for modes 1, 2, 3 and 4
Battery percentage
Brushing Timer, after what toothbrush will turn off

matissrapsa committed:
Update sensor.py

matissrapsa committed:
Update laifen.py

UrbanTechIO commented:
I have tested the changes you made, looks great, however, i think it is better to keep Oscillatinon speed, vibration and range and change their values according to the mode the brush is using.
also Brushing timer shows fixed value, I think it is bettter to use that value and count down to 0, in this case someone can use that to trigger something or register for example (Complete Brushing) in a calendar. and remove the timer sensor since it is not valid anymore

Suggestions

1- remove Laifen Toothbrush Oscillation Range Mode 1 2 3 4 and change the value of Oscillation sensor to the picked mode
2- same for the other sensors
3- Brushing Timer sensor value is counting down to 0 while status is running

matissrapsa committed:
Update sensor.py

matissrapsa committed:
Update laifen.py

matissrapsa commented:
Step 1 done, now the Oscillation Range, Oscillation Speed, Vibration Strength are selected based on selected mode
About point 3, I didn't touch your timer, I added a brushing timer that is pulled from the toothbrush, later this is planned to be a dropdown where you can select anywhere from 1 to 5 min after what toothbrush will power off

Also added a filter where the data is only pulled from the correct string 52 characters long

UrbanTechIO commented:
I was doing some work on your version, i removed the Timer entity, and used your brushing timer counting down, i'll share with you the code, and check yours

async def async_update(self, *args):
        """Update the sensor state."""
        await self.coordinator.async_request_refresh()
        if self.device.result is not None:
            status = self.device.result.get("status")
            if status == "Running" and self._timer_task is None:
                self._timer_state = self.device.result.get("brushing_timer")  # Reset timer when starting
                self._timer_task = asyncio.create_task(self._run_timer())
            elif status == "Idle" and self._timer_task is not None:
                self._timer_task.cancel()
                self._timer_task = None
                self._timer_state = self.device.result.get("brushing_timer")  # Reset timer when stopped
